arangod/Aql/OptimizerRules.cpp

```diff
@@ -3903,13 +3903,261 @@ void arangodb::aql::scatterInClusterRule(Optimizer* opt, std::unique_ptr<Executi
+                                      createKeys, allowSpecifiedKeys);
+  TRI_ASSERT(distNode != nullptr);
+  return distNode;
+}
+
+// Create a new GatherNode for the DistributeNode passed in node, and
+// register it with the plan
+//
+// TODO: Really Scatter/Gather and Distribute/Gather should be created in pairs.
+auto arangodb::aql::createGatherNodeFor(ExecutionPlan& plan, DistributeNode* node)
+    -> GatherNode* {
+  auto const collection = node->collection();
+
+  auto const sortMode = GatherNode::evaluateSortMode(collection->numberOfShards());
+  auto const parallelism = GatherNode::Parallelism::Undefined;
+  return plan.createNode<GatherNode>(&plan, plan.nextId(), sortMode, parallelism);
+}
+
+//
+// for a node `at` of type
+//  - INSERT, REMOVE, UPDATE, REPLACE, UPSERT
+//  - TRAVERSAL, SHORTEST_PATH, K_SHORTEST_PATHS,
+// we transform
+//
+// parents[0] -> `node` -> deps[0]
+//
+// into
+//
+// parents[0] -> GATHER -> REMOTE -> `node` -> REMOTE -> DISTRIBUTE -> deps[0]
+//
+// We can only handle the above mentioned node types, because the setup of
+// distribute and gather requires knowledge from these nodes.
+//
+// Note that parents[0] might be `nullptr` if `node` is the root of the plan,
+// and we handle this case in here as well by resetting the root to the
+// inserted GATHER node.
+//
+auto arangodb::aql::insertDistributeGatherSnippet(ExecutionPlan& plan,
+                                                  ExecutionNode* at, SubqueryNode* snode)
+    -> DistributeNode* {
+  auto const parents = at->getParents();
+  auto const deps = at->getDependencies();
+
+  // This transforms `parents[0] -> node -> deps[0]` into `parents[0] -> deps[0]`
+  plan.unlinkNode(at, true);
+
+  // create, and register a distribute node
+  DistributeNode* distNode = createDistributeNodeFor(plan, at);
+  TRI_ASSERT(distNode != nullptr);
+  distNode->addDependency(deps[0]);
+
+  // TODO: This dance is only needed to extract vocbase for
+  //       creating the remote node. The vocbase parameter for
+  //       the remote node does not seem to be really needed, since
+  //       the vocbase is stored in plan (and this variable is actually used in)
+  //       some code, so maybe this parameter could be removed?
+  auto const* collection = distNode->collection();
+  TRI_vocbase_t* vocbase = collection->vocbase();
+
+  // insert a remote node
+  ExecutionNode* remoteNode =
+      plan.createNode<RemoteNode>(&plan, plan.nextId(), vocbase, "", "", "");
+  remoteNode->addDependency(distNode);
+
+  // re-link with the remote node
+  at->addDependency(remoteNode);
+
+  // insert another remote node
+  remoteNode =
+      plan.createNode<RemoteNode>(&plan, plan.nextId(), vocbase, "", "", "");
+  remoteNode->addDependency(at);
+
+  // insert a gather node matching the distribute node
+  auto* gatherNode = createGatherNodeFor(plan, distNode);
+  gatherNode->addDependency(remoteNode);
+
+  // Song and dance to deal with at being the root of a plan or a subquery
```

Comment:
```suggestion
TRI_ASSERT(parents.size() < 2);
  // Song and dance to deal with at being the root of a plan or a subquery
```

This code only works if do not have multiple parents of this node.
I think right now, we do not create such kind of plans here, but this is future proof